compiler_gym/datasets/datasets.py

```diff
@@ -0,0 +1,480 @@
+       of datasets from their physical installation.
+
+    For the first goal, methods such as :meth:`benchmark()` provide a convenient
+    interface for randomly selecting benchmarks across multiple datasets:
+
+        >>> for i in range(100):
+        ...     benchmark = env.datasets.benchmark()
+
+    To iterate over a specific dataset, the datasets object can be indexed:
+
+        >>> for i in range(100):
+        ...     benchmark = env.datasets["cbench-v1"].benchmark()
+
+    The datasets object can be also be iterated over to list the available
+    datasets:
+
+        >>> for dataset in env.datasets:
+        ...     print(dataset.name)
+        benchmark://cbench-v1
+        benchmark://github-v0
+        benchmark://npb-v0
+
+    For the second goal, the datasets interface adds an *active* vs *installed*
+    datasets abstraction. An active dataset will be included in the
+    :meth:`datasets() <compiler_gym.datasets.Datasets.datasets>`,
+    :meth:`benchmarks() <compiler_gym.datasets.Datasets.benchmarks>`, and
+    :meth:`benchmark_uris() <compiler_gym.datasets.Datasets.benchmark_uris>`
+    value iterators. Upon first use, it will be installed, which may require an
+    expensive process of downloading remote resources and caching them locally.
+
+    Use :meth:`deactivate() <compiler_gym.datasets.Datasets.deactivate>` to mark
+    a dataset as inactive such that it won't be included in set of available
+    datasets, but without removing any files that would require a lengthy
+    re-installation:
+
+        >>> env.datasets.install("github-v0")
+        >>> env.datasets.deactivate("github-v0")
+        >>> for dataset in env.datasets:
+        ...     print(dataset.name)
+        benchmark://cbench-v1
+        benchmark://npb-v0
+        >>> env.datasets["github-v0"].installed
+        True
+
+    The :meth:`activate() <compiler_gym.datasets.Datasets.activate>` method can
+    later be used to mark this dataset as ready for use again, without needing
+    to reinstall anything:
+
+        >>> env.datasets.activate("github-v0")
+        >>> for dataset in env.datasets:
+        ...     print(dataset.name)
+        benchmark://cbench-v1
+        benchmark://github-v0
+        benchmark://npb-v0
+
+    The installation of datasets occurs lazily and automatically, however there
+    are :meth:`install() <compiler_gym.datasets.Datasets.install>` and
+    :meth:`uninstall() <compiler_gym.datasets.Datasets.uninstall>` methods
+    should you wish to perform this manually.
+    """
+
+    def __init__(
+        self,
+        datasets: Iterable[Dataset],
+        random: Optional[np.random.Generator] = None,
+    ):
+        # A mapping from dataset name to the dataset instance.
+        self._datasets: Dict[str, Dataset] = {d.name: d for d in datasets}
+        # The set of activate datasets. All datasets except hidden ones are
+        # initialized as activate.
+        self._active_datasets: Set[str] = set(
```

Comment:
I don't get the need for this.

Reply:
v2 rewrite: Renamed to `_visible_datasets`. I still need to keep track of the subset of `self._datasets` that can be used for iterating over